src/addr.rs

```diff
@@ -290,6 +311,17 @@ impl SubAssign<u64> for VirtAddr {
     }
 }
 
+impl CheckedSub<u64> for VirtAddr {
+    type Output = Self;
+
+    fn checked_sub(self, rhs: u64) -> Option<Self::Output> {
+        self.0
+            .checked_sub(rhs)
+            .and_then(is_canonical)
+            .map(Self::new)
```

Comment:
Similarly, this can just be:
```rust
from_canonical(self.0.checked_sub(rhs)?)
```
Ideally it would be:
```rust
VirtAddr::try_new(self.0.checked_sub(rhs)?).ok()
```
but that doesn't work see #299

Reply:
Using your above suggestion of `from_canonical` I ended up just cleaning up the chain a bit.

```
    fn checked_add(self, rhs: u64) -> Option<Self::Output> {
        self.0.checked_add(rhs).and_then(from_canonical)
    }
``` 

If you really prefer, i can switch to the error propagation operator.

Reply:
`and_then` is fine, I actually like it better than my proposed solution.